### PR TITLE
feat(prefer-const): add rule

### PR DESCRIPTION
**Pull request by mikededo**
Adds a new rule named `prefer-const` which:

- Overrides the ESLint `prefer-const rule to support Svelte syntax, as discussed
  in #818.
- ~Ports~ Proxies the original `prefer-const` rule from ESLint, adapting it for Svelte.

~This PR is an updated version of #816, which did not include types. Also, some
opinionated changes were made in terms of code style, not in terms of behavior.~
Avoided re-implemeting the core rule, by using the `getCoreRule` and proxying the
default implementation.

I found that this rule could be very useful for Svelte, and there could be many
developers who would benefit from it. ~At the same time, I thought that the
initial implementation was a good starting point, and that adding types and
would make the rule more robust to changes.~  
However, the rule should not have many changes, as how variables are handled in
JS is not likely to change.

> Acknowledgement: this rule is based on the `prefer-const` rule from ESLint and
> the initial work done by @bfanger.

### CODE REVIEW

**Comment by ota-meshi** [comment]
Thanks for working on implementing this rule!
However, could you please implement it using `getCoreRule()` in the same way as `no-inner-declarations`?
https://github.com/sveltejs/eslint-plugin-svelte/blob/main/packages/eslint-plugin-svelte/src/rules/no-inner-declarations.ts

I think it just avoids calling listeners on `VariableDeclaration` that use runes.

**Comment by mikededo** [comment]
Thanks @ota-meshi for pointing this out, my initial idea was to use a proxy for the rule, to avoid having a complete re-write. Should've asked before spending so many hours 😅 🤦🏼 

Will look into it. Maybe I'll end up closing this PR in favour of another one.

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import { defineWrapperListener, getCoreRule } from '../utils/eslint-core';
+
+const coreRule = getCoreRule('prefer-const');
+
+type Declaration = TSESTree.Expression;
+
+/**
+ * Finds and returns the callee of a declaration node within variable declarations or object patterns.
+ */
+function findDeclarationCallee(node: Declaration) {
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:13]: I don't think we need a type alias.

```suggestion
function findDeclarationCallee(node: TSESTree.Expression) {
```

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	return null;
+}
+
+/**
+ * Determines if a declaration should be skipped in the const preference analysis.
+ * Specifically checks for Svelte's state management utilities ($state, $props, $derived).
+ */
+function shouldSkipDeclaration(declaration: Declaration | null) {
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:35]:
```suggestion
function shouldSkipDeclaration(declaration: TSESTree.Expression | null) {
```

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Finds and returns the callee of a declaration node within variable declarations or object patterns.
+ */
+function findDeclarationCallee(node: Declaration) {
+	const { parent } = node;
+	if (parent.type === 'VariableDeclarator' && parent.init?.type === 'CallExpression') {
+		return parent.init.callee;
+	}
+
+	if (
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:19]: What code passes through this branch?

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	if (!declaration || !declaration.parent) {
+		return false;
+	}
+
+	const callee = findDeclarationCallee(declaration);
+	if (!callee) {
+		return false;
+	}
+
+	if (callee.type === 'Identifier' && ['$state', '$props', '$derived'].includes(callee.name)) {
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:45]: Why did you exclude `$state` from rule checking?
I think it's ok to have a `const` variable as long as the user doesn't reassign it in their code 🤔

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		callee.object.name === '$derived' &&
+		callee.property.type === 'Identifier' &&
+		callee.property.name === 'by'
+	) {
+		return true;
+	}
+	if (
+		callee.object.name === '$state' &&
+		callee.property.type === 'Identifier' &&
+		callee.property.name === 'frozen'
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:63]: I think it's `$state.raw`.

```suggestion
		callee.property.name === 'raw'
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]
Thank you for this PR!

File: packages/eslint-plugin-svelte/package.json [hunk 6]
```diff
@@ -30,7 +30,7 @@
     "cover": "nyc --reporter=lcov pnpm run test",
     "debug": "pnpm run mocha \"tests/src/**/*.ts\" --reporter dot --timeout 60000",
     "lint": "run-p lint:*",
-    "lint-fix": "pnpm run lint:es --fix && pnpm run lint:style --fix",
+    "lint:fix": "pnpm run lint:es --fix",
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/package.json:33]: Can you revert the changes?

Comment (ota-meshi) [reply packages/eslint-plugin-svelte/package.json:33]: Oh, `lint:style` doesn't exist now. Could you revert the name of script?

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts (hunk 4 above)
Comment (mikededo) [reply packages/eslint-plugin-svelte/src/rules/prefer-const.ts:45]: This was part of the discussion in #818. I also proposed having an option that allows the user to decide [which reactive statements](https://github.com/sveltejs/eslint-plugin-svelte/issues/818#issuecomment-2505770979) should be excluded from this rule.

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts (hunk 3 above)
Comment (mikededo) [reply packages/eslint-plugin-svelte/src/rules/prefer-const.ts:19]: After investigating for a bit, I realised that there's no valid JavaScript/TypeScript syntax that would create an AST structure, taking into account that the node that's passed into `findDeclarationCallee` comes from a `VariableDeclaration.declarations[number].init`.  The structure doesn't make sense because a `Property` node’s parent cannot simultaneously be a `VariableDeclarator`. 

I'll delete this branch.

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts (hunk 4 above)
Comment (ota-meshi) [reply packages/eslint-plugin-svelte/src/rules/prefer-const.ts:34]: I read the thread. So it's ok for you to use `const` for `$state` without reassignment, right?
I'm also ok with `const`, so I think it's better to remove the `$state` branch from the PR and check `$state`.

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import { createRule } from '../utils';
+import type { TSESTree } from '@typescript-eslint/types';
+
+import { defineWrapperListener, getCoreRule } from '../utils/eslint-core';
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:4]: I changed this package to ESM.
https://github.com/sveltejs/eslint-plugin-svelte/pull/932

Could you please merge the main branch and fix it by adding `.js` to the import statement of this PR?

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts (hunk 4 above)
Comment (mikededo) [reply packages/eslint-plugin-svelte/src/rules/prefer-const.ts:34]: Yes!

Also, what about the option to customize which reactive assignments to ignore?

Comment (ota-meshi) [reply packages/eslint-plugin-svelte/src/rules/prefer-const.ts:34]: > Also, what about the option to customize which reactive assignments to ignore?

I don't think there's a need for an option for now. If someone needs one, I think they'll suggest it in an issue. If necessary, we'll add the option then.

**Comment by mikededo** [comment]
Hey @ota-meshi, would it be possible to have another review when possible? 🎅🏼

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+	if (
+		callee.object.name === '$derived' &&
+		callee.property.type === 'Identifier' &&
+		callee.property.name === 'by'
+	) {
+		return true;
+	}
+	if (
+		callee.object.name === '$state' &&
```
Comment (ota-meshi) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:50]: I think there is probably no need to ignore `$state.raw` from the check for now.


https://github.com/sveltejs/eslint-plugin-svelte/pull/933#discussion_r1865388045

**Review by ota-meshi — CHANGES_REQUESTED** [review]
Sorry for the late reply 🙇
It almost looks good to me, but I have one comment.

File: packages/eslint-plugin-svelte/src/rules/prefer-const.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	if (parent.type === 'VariableDeclarator' && parent.init?.type === 'CallExpression') {
+		return parent.init.callee;
+	}
+
+	return null;
+}
+
+/**
+ * Determines if a declaration should be skipped in the const preference analysis.
+ * Specifically checks for Svelte's state management utilities ($state, $props, $derived).
```
Comment (baseballyama) [inline packages/eslint-plugin-svelte/src/rules/prefer-const.ts:22]:
```suggestion
 * Specifically checks for Svelte's state management utilities ($props, $derived).
```

I think here doesn't handle `$state`.

**Review by baseballyama — APPROVED** [review]
